### PR TITLE
refactor: notUndefined instanciate the error when necessary

### PR DESCRIPTION
**Pull request by Defferrard**

### CODE REVIEW

File: src/utils/assertions.test.ts [hunk 1]
```diff
@@ -1,30 +1,29 @@
-import { notUndefined } from './assertions';
+import { asNonNull } from './assertions';
 
 describe('notUndefined', () => {
```
Comment (spaenleh) [inline src/utils/assertions.test.ts:3]: Update test name.

File: src/utils/assertions.test.ts
Comment (spaenleh) [inline src/utils/assertions.test.ts:1]: Add tests for the `isNonNull` function ?

**Review by spaenleh — APPROVED** [review]
Nice !

File: src/utils/assertions.ts [hunk 2]
```diff
@@ -1,16 +1,31 @@
 import { UnexpectedError } from './errors';
 
+export type Nullable<T> = T | null | undefined;
+
+/**
+ *  Returns true if the object is not null or undefined.
+ * @param object Object to check
+ * @returns True if the object is not null or undefined, false otherwise
+ */
+export function isNonNull<T>(object: Nullable<T>): object is T {
```
Comment (ReidyT) [inline src/utils/assertions.ts:10]: What about this suggestion?

```suggestion
export function isDefined<T>(object: Nullable<T>): object is T {
```

File: src/utils/assertions.ts [hunk 3]
```diff
@@ -1,16 +1,31 @@
...
+ * Throw an error if the object is undefined or null. Otherwise, return the object in a not-undefined type.
  * @param object Object to check
- * @param error Error to throw if the object is undefined
+ * @param error Error to throw if the object is undefined or null
  * @returns Object in a not-undefined type
  */
-export function notUndefined<T>(object: T | undefined, error?: Error): T {
-  if (object === undefined) {
-    throw error || new UnexpectedError();
+export function asNonNull<T, Err extends Error, Args extends unknown[]>(
```
Comment (ReidyT) [inline src/utils/assertions.ts:20]: What do you think about:

```suggestion
export function assertIsDefined<T, Err extends Error, Args extends unknown[]>(
```

`asNonNull` doesn't really reflect the throw if null. Also, I'm wondering if we could not merge it with `assertNonNull`?

**Review by ReidyT — APPROVED** [review]
Nice PR! LGTM, I just let you some comments 🤖

**Pull request merged**